src/utils/mac-table.h

```diff
@@ -224,102 +274,63 @@ static inline void *pg_mac_table_ptr_get(struct pg_mac_table *ma,
 	return ma->ptrs[part1]->entries[part2];
 }
 
-static inline void pg_mac_table_iterator_next(struct pg_mac_table_iterator *it)
-{
-	struct pg_mac_table *ma = it->table;
-	uint32_t part1 = pg_mac_table_part1(it->key);
-	uint32_t part2 = pg_mac_table_part2(it->key);
-
-	it->key.part2 = part2 + 1;
-	part2 = pg_mac_table_part2(it->key);
-	if (!part2) {
-		it->key.bytes32[0] = part1 + 1;
-		if (!pg_mac_table_part1(it->key)) {
-			it->is_end = 1;
-			return;
-		}
-		part1 = pg_mac_table_part1(it->key);
-	}
-
-	uint64_t unset = (~0LLU) << (pg_mac_table_mask_pos(part1));
-	uint64_t sub_unset= (~0LLU) << (pg_mac_table_mask_pos(part2));
-	int first_mask = pg_mac_table_mask_idx(part1);
-	int first_sub_mask = pg_mac_table_mask_idx(part2);
-	uint64_t *masks = ma->mask;
-
-	if (masks[first_mask] && ctz64(masks[first_mask]) > ctz64(unset))
-		sub_unset = (~0LLU);
-	for (int i = first_mask; i < PG_MAC_TABLE_MASK_SIZE; ++i) {
-		if (!masks[i])
-			continue;
-		PG_FOREACH_BIT(masks[i] & unset, i2) {
-			uint64_t *sub_masks = ma->ptrs[(i * 64) + i2]->mask;
-
-			for (int j = first_sub_mask;
-			     j < PG_MAC_TABLE_MASK_SIZE; ++j) {
+#define MAC_TABLE_IT_NEXT pg_mac_table_iterator_e_next
+#define MAC_TABLE_TYPE elem
+#define MAC_TABLE_SETTER(inner_t, useless) (it->c_ptr = (inner_t)->entries);
 
-				if (!sub_masks[j])
-					continue;
+#include "mac-table-it-next.h"
 
-				PG_FOREACH_BIT(sub_masks[j] & sub_unset, j2) {
-					it->key.bytes32[0] = (i * 64) + i2;
-					it->key.part2 = (j * 64) + j2;
-					return;
-				}
-				sub_unset = ~0LLU;
-			}
-			first_sub_mask = 0;
-		}
-		unset = ~0LLU;
-	}
-	it->is_end = 1;
-}
+#define MAC_TABLE_SETTER(inner_t, idx) (it->v_ptr = (inner_t)->entries[idx]);
+#define MAC_TABLE_TYPE ptr
+#define MAC_TABLE_IT_NEXT pg_mac_table_iterator_next
 
-static inline void pg_mac_table_iterator_init(struct pg_mac_table_iterator *it,
-					      struct pg_mac_table *tbl)
-{
-	it->key.mac = 0;
-	it->is_end = 0;
-	it->table = tbl;
+#include "mac-table-it-next.h"
 
```

Comment:
double include of  "mac-table-it-next.h"

Reply:
yes